### DISCUSSION TITLE
Matrix operations and the associated statistical functions derived from them

### DISCUSSION THREAD

**Discussion by Peter-Jacob** [adesutherland/CREXX#416] [Show and tell]
I ended up re-coding my thesis—not in Fortran IV, and it didn’t take 6 months this time. Just 2 days, with a little help from my friend (that friend!) 

# Matrix Operations Plugin for CREXX/PA

## Overview
This plugin provides comprehensive matrix operations for the CREXX/PA system, including linear algebra and statistical functions.

## Memory Management
- Cross-platform heap storage using MATRIX_ALLOC/MATRIX_FREE macros
- Windows: HeapAlloc/HeapFree from process heap
- Unix/Linux/macOS: malloc/free from C standard library
- Maximum matrices: 100
- Automatic cleanup on matrix free or plugin unload

### Basic Operations
- `mmultiply(m0, m1, 'description')` - Matrix multiplication (m0 × m1 → 'description')
- `mprod(m0, scalar, 'description')` - Scalar multiplication (m0 × scalar → 'description')
- `mtranspose(m0, 'description')` - Matrix transpose (m0 → 'description')
- `minvert(m0, 'description')` - Matrix inversion (m0 → 'description')

### Matrix Creation and Management
- `mcreate(rows, cols, id)` - Create new matrix
- `mset(m0, row, col, value)` - Set matrix element
- `mprint(m0)` - Print matrix
- `mfree(m0)` - Free matrix (if m0 < 0, frees all)

### Linear Algebra
- `mdet(m0)` - Calculate determinant
- `mlu(m0, L, U)` - LU decomposition
- `mqr(m0, Q, R)` - QR decomposition
- `meigen(m0, vec)` - Calculate dominant eigenvalue/vector
- `mrank(m0)` - Calculate matrix rank

### Statistical Operations
- `mcov(m0, 'description')` - Covariance matrix
- `mcorr(m0, 'description')` - Correlation matrix
- `mmean(m0, axis, 'description')` - Row/column means (axis: 0=rows, 1=cols)
- `mstandard(m0, 'description')` - Standardize matrix columns

## Performance Features
- Cache-friendly blocking for matrix operations
- Block size optimization (MATRIX_BLOCK_SIZE = 64)
- Loop unrolling for better instruction pipelining
- Efficient memory access patterns
- Numerical stability checks

## Error Handling
- Input validation
- Memory allocation checks
- Numerical stability checks
- Detailed error codes

- **mtranspose(m, id)** → matrix_number
  - Transposes a matrix
  - Parameters:
    - m: Matrix number
    - id: Result matrix identifier
  - Returns: New matrix number or error code

- **mstandard(m, id)** → matrix_number
  - Standardizes matrix columns (z-score transformation)
  - Parameters:
    - m: Matrix number
    - id: Result matrix identifier
  - Returns: New matrix number or error code

### Matrix Element Access
- **mset(m, row, col, value)** → status
  - Sets a matrix element value
  - Parameters:
    - m: Matrix number
    - row: Row number (1-based)
    - col: Column number (1-based)
    - value: New value (float)
  - Returns: 0 on success, error code on failure

- **mprint(m)** → status
  - Prints matrix contents
  - Parameters:
    - m: Matrix number
  - Returns: 0 on success, error code on failure

## Error Codes
- MATRIX_SUCCESS (0): Operation successful
- MATRIX_INVALID_PARAM (-1): Invalid parameters
- MATRIX_NO_SLOTS (-12): No available matrix slots
- MATRIX_ALLOC_CB (-16): Control block allocation failed
- MATRIX_ALLOC_DATA (-20): Data allocation failed
- MATRIX_NULL (-30): Matrix is NULL
- MATRIX_INVALID_INDEX (-31): Invalid matrix index
- MATRIX_CORRUPT (-32): Matrix structure is corrupt

## Example Usage
```
/ Create and initialize a matrix /
m1 = mcreate(3, 3, "Test Matrix")
call mset m1, 1, 1, 1.0
call mset m1, 2, 2, 1.0
call mset m1, 3, 3, 1.0
/ Print the matrix /
call mprint m1
/ Create transpose /
m2 = mtranspose(m1, "Transposed")
call mprint m2
/ Multiply by scalar /
m3 = mprod(m1, 2.0, "Doubled")
call mprint m3
/ Free resources /
call mfree m1
call mfree m2
call mfree m3
```
```
rexx
/ Create and multiply two matrices /
m1 = call mcreate(3, 3, "Matrix1")
m2 = call mcreate(3, 3, "Matrix2")
call mset(m1, 1, 1, 2.0) / Set element at row 1, col 1 /
result = call mmultiply(m1, m2, "Result")
call mprint(result)
call mfree(result)
```

## Implementation Notes
- Matrix indices are 1-based in REXX calls
- All matrices are stored in double precision
- Maximum number of matrices: 100
- Memory is managed through Windows HeapAlloc otherwise it is heap storage (calloc)
- Matrix operations validate inputs before processing

## Error Handling
All functions return error codes for invalid operations. Check return values for:
- Invalid matrix numbers
- Memory allocation failures
- Invalid dimensions
- Corrupt matrix structures

## Return Codes
- 0: Success
- -1: Invalid parameter
- -2: Memory allocation failure
- -3: Singular matrix
- -4: Dimension mismatch
- -8: Memory free error

## Notes
- All indices are 1-based for REXX compatibility
- Matrices are stored in row-major order
- Double precision floating point used throughout

### Operation Status Codes
- MATRIX_SUCCESS (0): Operation completed successfully
- MATRIX_INVALID_PARAM (-1): Invalid input parameters
  - Matrix dimensions don't match
  - Invalid matrix number
  - NULL matrix identifier
  - Invalid row/column indices

### Memory Related Errors
- MATRIX_NO_SLOTS (-12): No available matrix slots
  - Maximum number of matrices (100) reached
  - Need to free some matrices before creating new ones
- MATRIX_ALLOC_CB (-16): Control block allocation failed
  - System memory allocation error
  - Not enough memory for matrix structure
- MATRIX_ALLOC_DATA (-20): Data allocation failed
  - Not enough memory for matrix data
  - System memory allocation error

### Matrix Validation Errors
- MATRIX_NULL (-30): Matrix is NULL
  - Attempting to operate on a NULL matrix
  - Matrix may have been freed
- MATRIX_INVALID_INDEX (-31): Invalid matrix index
  - Matrix number out of range (0-99)
  - Negative matrix number
- MATRIX_CORRUPT (-32): Matrix structure is corrupt
  - Invalid internal pointers
  - Memory corruption detected

### Operation-Specific Errors
- MATRIX_SINGULAR (-40): Matrix is singular
  - Determinant is zero
  - Matrix cannot be inverted
- MATRIX_DIM_MISMATCH (-41): Matrix dimensions mismatch
  - Incompatible dimensions for operation
  - Example: multiplication of incompatible matrices
- MATRIX_NOT_SQUARE (-42): Matrix must be square
  - Operation requires square matrix
  - Example: LU decomposition
- MATRIX_NO_CONVERGENCE (-43): Algorithm did not converge
  - Maximum iterations reached
  - Example: eigenvalue calculation

### Memory Management Errors
- MATRIX_FREE_ERROR (-8): Error freeing matrix memory
  - System memory deallocation error
  - Invalid memory block
- MATRIX_FREE_CB_ERROR (-12): Error freeing control block
  - System memory deallocation error
  - Invalid control block pointer

### Linear Algebra Examples
rexx
/ LU Decomposition /
m = call mcreate(2, 2, "Original")
call mset(m, 1, 1, 4.0)
call mset(m, 1, 2, 3.0)
call mset(m, 2, 1, 6.0)
call mset(m, 2, 2, 3.0)
L = call mlu(m, "L", "U") / Returns L, U is L+1 /
call mprint(L)
call mprint(L+1)
/ Matrix Rank /
rank = call mrank(m)
say "Matrix rank:" rank
/ Eigenvalue /
ev = call meigen(m, "eigenvector")
call mprint(ev) / First element is eigenvalue, rest is eigenvector /

### Statistical Analysis
rexx:call md
/ Create data matrix /
data = call mcreate(4, 3, "DataSet")
/ Row 1 /
call mset(data, 1, 1, 1.2)
call mset(data, 1, 2, 2.3)
call mset(data, 1, 3, 3.4)
/ Row 2 /
call mset(data, 2, 1, 2.3)
call mset(data, 2, 2, 3.4)
call mset(data, 2, 3, 4.5)
/ Row 3 /
call mset(data, 3, 1, 3.4)
call mset(data, 3, 2, 4.5)
call mset(data, 3, 3, 5.6)
/ Row 4 /
call mset(data, 4, 1, 4.5)
call mset(data, 4, 2, 5.6)
call mset(data, 4, 3, 6.7)
/ Calculate statistics /
means = call mmean(data, 1, "ColumnMeans") / Column means /
covar = call mcov(data, "CovMatrix") / Covariance matrix /
corr = call mcorr(data, "CorrMatrix") / Correlation matrix /
call mprint(means)
call mprint(covar)
call mprint(corr)

**Comment by Peter-Jacob** [comment]
A new version loaded with additional functionality:
---

# Matrix Operations Library

---

## Introduction
The **Matrix Operations Library** is a comprehensive and versatile toolkit rooted in the principles of **Linear Algebra** and enriched with robust **statistical capabilities**. Linear Algebra forms the foundation for understanding and manipulating matrices and vector spaces, while the statistical components extend its application to data analysis and interpretation. This library is designed to facilitate efficient matrix computations, statistical analysis, and data visualization.

With its ability to handle both fundamental operations, such as matrix multiplication and inversion, and advanced statistical tasks, like covariance and factor analysis, the library bridges the gap between mathematical theory and practical data-driven applications.

## Overview
### Summary
This library provides:
- **Basic Matrix Functions**: Includes creation, multiplication, inversion, transposition, and arithmetic operations.
- **Matrix Characteristics**: Tools for calculating determinants, ranks, and decompositions (e.g., LU, QR, and SVD).
- **Statistical Analysis**: Functions for covariance, correlation, mean, standard deviation, and advanced metrics like skewness and kurtosis.
- **Factor Analysis**: Includes support for rotation methods (Varimax, Promax) and factor score calculations.
- **Visualization Tools**: Enables matrix data visualization through line plots, scatterplots, and ASCII-based heatmaps.
- **Error Handling**: Comprehensive mechanisms to handle common matrix errors, ensuring robust and fail-safe operations.
- **Performance Optimization**: Features for parallel processing, memory optimization, and numerical stability.

---

## Basic Matrix Functions

### Matrix Creation

### Creating a New Matrix
- **Purpose**: Creates a new matrix with specified dimensions.
- **Parameters**:
    - `rows`: Number of rows (int)
    - `cols`: Number of columns (int)
    - `id`: Matrix identifier (string)
- **Returns**: Matrix handle (int)
```c
id = mcreate(3, 3, "myMatrix");
```

### Matrix Multiplication (`mmult`,'description')
- **Purpose**: Multiplies two matrices: \( C = A * B \)
- **Parameters**:
    - `m0`: First input matrix (int)
    - `m1`: Second input matrix (int)
    - `mid`: Result matrix identifier (string)
- **Returns**: new Matrix handle (int)

### Scalar Multiplication (`mprod`,'description')
- **Purpose**: Multiplies a matrix by a scalar: \( B = A * scalar \)
- **Parameters**:
    - `m0`: Input matrix (int)
    - `prod`: Scalar value (float)
    - `mid`: Result matrix identifier (string)
- **Returns**: new Matrix handle (int)


### Matrix Inversion (`minvert`,'description')
- **Purpose**: Computes the inverse of a matrix: \( B = A^{-1} \)
- **Parameters**:
    - `m0`: Input matrix (int)
    - `mid`: Result matrix identifier (string)
- **Returns**: new Matrix handle (int)

### Matrix Transposition (`mtranspose`,'description')
- **Purpose**: Computes the transpose of a matrix: \( B = A^T \)
- **Parameters**:
    - `m0`: Input matrix (int)
    - `mid`: Result matrix identifier (string)
- **Returns**: Matrix handle (int)

### Matrix Arithmetic Examples
```c
madd(m1, m2, "result"); // Addition
mmult(m1, m2, "result"); // Multiplication
mtrans(m1, "result"); // Transpose
```

### Matrix Characteristics
### `mdet`
- **Purpose**: Calculate matrix determinant
- **Parameters**:
  - m0: Input matrix (matrix-handle)
- **Returns**: Success status (integer)

### `mrank`
- **Purpose**: Calculate matrix rank
- **Parameters**:
  - m0: Input matrix (matrix-handle)
- **Returns**: Matrix rank (integer)
### `mlu`
- **LU Decomposition**: `mlu(m0, L, U)` — Performs LU decomposition, returning lower (`L`) and upper (`U`) triangular matrices.
- m0: Input matrix (matrix-handle)
### `mprint`
- **Purpose**: Print matrix to output
- **Parameters**:
  - m0: Matrix to print (matrix-handle,"alternative-heading") - if empty Matrix heading defined with MCREATE is used
- **Returns**: Success status (integer)

---

## Statistical Functions

### Basic Statistical Operations

#### Covariance (`mcov`,'description')
Covariance is a statistical measure that quantifies the extent to which two variables change together. It provides insight into the direction of the linear relationship between variables:

Positive Covariance: Indicates that as one variable increases, the other tends to increase as well.
Negative Covariance: Suggests that as one variable increases, the other tends to decrease.

- **Purpose**: Computes the covariance matrix.
- **Returns**: Covariance matrix handle (int)
```c
cov_id = mcov(id, "Covariance");
```

#### Correlation (`mcorr`,'description')
Correlation is a statistical measure that quantifies the strength and direction of the linear relationship between two variables. Unlike covariance, correlation is standardized, providing values within the range
− −1 to 1

- 1: Perfect positive correlation (variables increase together).
- -1: Perfect negative correlation (one variable increases as the other decreases).
- 0: No linear relationship between the variables.


- **Purpose**: Computes the correlation matrix.
- **Returns**: Correlation matrix handle (int)
```c
corr_id = mcorr(id, "Correlation");
```

### Mean and Standard Deviation
Mean and Standard Deviation (StdDev) are fundamental statistical measures used to describe and analyze datasets:

**Mean**:

Represents the average value of a dataset.
Calculated by summing all values and dividing by the number of observations.
Reflects the central tendency of the data.

**Standard Deviation**:

Measures the dispersion or spread of data around the mean.
A smaller standard deviation indicates that data points are closer to the mean, while a larger standard deviation signifies greater variability.

### `mmean(m0, row)`
- **Purpose**: Calculate mean along specified columns
- **Parameters**:
  - m0: Input matrix (matrix-handle)
  - column number
- **Returns**: Mean value (.float)
 
### `mstdev(m0, row)`
- **Purpose**: Calculates standard deviation along specified columns
- **Parameters**:
  - m0: Input matrix (matrix-handle)
  - column number
- **Returns**: standard deviation value (.float)

```c
mean = mmean(id, 1); // Row 1 mean
stddev = mstdev(id, 2); // Row 2 standard deviation
```
### Matrix Helper Functions
### `colstats`
- **Purpose**: General statistical analysis
- **Parameters**:
  - m0: Input matrix (matrix-handle)
  - mode: row number 
  - **Returns**: Vector - Matrix(4,1)  with input matrix details
```
    row 1: StdDev
    row 2: median
    row 3: skewness
    row 4: kurtosis 
```
 
### Factor Analysis
Factor Analysis is a statistical method designed to uncover the underlying structure in a dataset by reducing observed variables into a smaller set of latent variables, called factors. These factors represent shared variance among the observed variables and are used to simplify data interpretation while retaining essential information. The technique is widely applied in psychology, social sciences, finance, and other fields for tasks such as dimensionality reduction and identifying relationships among variables.
Identify and group correlated variables into factors.
Reduce dimensionality while preserving meaningful patterns in the data.
### Factor Analysis (`mfactor`)
- **Purpose**: Performs factor analysis, including options for rotation (e.g., Varimax, Promax).
- **Parameters**:
    - `m0`: Input matrix (int)
    - `factors`: Number of factors (int)
    - `mid`: Result matrix description
    - `parms`: Additional options (string, e.g., rotation type)
- **Returns**: Matrix handle (int)
```c
loadings = mfactor(mfdata, 2, "Factor Analysis Varimax","Rotate=VARIMAX");
loadings2 = mfactor(mfdata, 1, "Factor Analysis Varimax","Rotate=Varimax,SCORE,diag=3")
loadings3 = mfactor(mfdata, 2, "Factor Analysis unrotated","Rotate=none,SCORES,diag=3")
loadings4 = mfactor(mfdata, 2, "Factor Analysis Quartimax","diag=3")
```

#### Factor Analysis Options
1. **ROTATE=type**: Type of rotation (e.g., `PROMAX`, `VARIMAX`, `NONE`).
2. **SCORE**: Enable factor score calculation.
3. **DIAG=value**: Diagnostic level.

### Advanced Matrix Operations
- **Specialized Decompositions**
  - Cholesky decomposition
  - QR decomposition
  - Singular Value Decomposition (SVD)
- **Matrix Functions**
  - Matrix exponential
  - Matrix logarithm
  - Matrix power
---

## Visualization

### Plotting Functions

#### `mplot`
- **Purpose**: Creates various plots like line, scatter, or bar charts.
- **Parameters**:
    - `m0`: Matrix ID (int)
    - `plot_type`: Plot type (string)
```c
mplot(id, "line"); // Line plot
```

#### `mfaplot`
- **Purpose**: Plots factor analysis results (e.g., scree plot, loadings).
- **Parameters**:
    - `m0`: Matrix ID (int)
    - `plot_type`: Plot type (string, e.g., `scree`, `loadings`)
```c
mfaplot(loadings, "scree"); // Scree plot
```

#### `masciiplot`
- **Purpose**: Creates ASCII-based plots (e.g., heat maps, histograms).
- **Parameters**:
    - `m0`: Matrix ID (int)
    - `plot_type`: Plot type (string)
```c
masciiplot(id, "heat"); // Heat map
```

---

## Error Handling

### Common Error Codes
- **General Errors**: `-1` for general errors, `-2` for invalid matrix ID.
- **Matrix Errors**: `-20` for dimension mismatch, `-21` for singular matrix.
- **Plotting Errors**: `-50` for invalid plot type, `-51` for missing Gnuplot.

### Example of Error Handling
```c
result = mmult(m1, m2, "result");
if (result < 0) {
    switch(result) {
        case -20: printf("Dimension mismatch\n"); break;
        case -5: printf("Memory allocation failed\n"); break;
    }
}
```

---

## Best Practices

1. **Always Check Return Values**: Ensure that matrix operations and functions return success before proceeding.
2. **Memory Management**: Clean up matrices with `mfree()` to prevent memory leaks.
3. **Error Recovery**: Implement proper error handling to manage issues such as dimension mismatches or singular matrices.

---

## Complete Example

```c
// Create a data matrix
data_id = mcreate(100, 5, "MyData");

// Calculate correlation matrix
corr_id = mcorr(data_id, "Correlations");

// Plot the correlation matrix as a heat map
masciiplot(corr_id, "heat");

// Perform factor analysis with 2 factors and Varimax rotation
loadings = mfactor(data_id, 2, "Factor Analysis Varimax", "Rotate=VARIMAX");

// Plot the factor analysis results
mfaplot(loadings, "scree");
```

## Extensions

### Matrix Types
- **Dense Matrix**: Standard row-major storage
- **Sparse Matrix**: Compressed storage for sparse data
- **Symmetric Matrix**: Optimized storage for symmetric data

### Error Categories
1. **Success (0)**
   - Successful operation completion

2. **Memory Management (-1 to -9)**
   - Memory allocation/deallocation
   - Handle management

3. **Dimension Related (-10 to -19)**
   - Size compatibility
   - Index bounds

4. **Calculation Related (-20 to -29)**
   - Numerical issues
   - Convergence problems

5. **Parameter Related (-30 to -39)**
   - Invalid inputs
   - Option parsing

6. **File Operations (-40 to -49)**
   - I/O errors
   - Format issues

7. **System Limits (-50 to -59)**
   - Resource constraints
   - System limitations
### Advanced Features
1. **Parallel Processing**
   - OpenMP support for large matrices
   - Thread-safe operations
   - Parallel algorithm implementations

2. **Memory Optimization**
   - Smart memory allocation
   - Cache-friendly algorithms
   - Memory pool for small matrices
   

### Performance Considerations
1. **Algorithm Selection**
   - Size-based optimization
   - Automatic method switching
   - Resource usage monitoring

2. **Cache Optimization**
   - Block operations
   - Stride optimization
   - Memory alignment

3. **Numerical Stability**
   - Condition number checking
   - Scaling strategies
   - Precision management

### Matrix Validation
- **Input Validation**
  - Dimension checks
  - Value range validation
  - Type compatibility
- **Result Validation**
  - Numerical stability checks
  - Convergence verification
  - Error bounds calculation

#### `matrix.stats`
- **Purpose**: General statistical analysis
- **Parameters**:
  - m0: Input matrix (matrix-handle)
  - mode: Analysis mode (string) - ROW or COL
- **Returns**: returns matrix dimension either column or row number 

#### `matrix.mplot`
- **Purpose**: Create matrix plot
- **Parameters**:
  - m0: Input matrix (matrix-handle)
  - plot_type: Type of plot (string)
- **Returns**: Success status (integer)

#### `masciiplot`
- **Purpose**: Create ASCII visualization
- **Parameters**:
  - m0: Input matrix (matrix-handle)
  - plot_type: Plot type (string)
- **Returns**: Success status (integer)

**Comment by adesutherland** [comment]
This is amazing! So a leading question if you had the required rexx objects, would you be willing to make a port to crexx itself?

**Reply by Peter-Jacob** [reply]
I developed the MATRIX functionality from scratch as a CREXX plugin; nothing was ported. 
Matrix arithmetic is computationally intensive, which is why I chose to implement it in C.
